dbms/src/Functions/FunctionsJson.h

```diff
@@ -530,11 +612,723 @@ class FunctionsJsonArray : public IFunction
+        size_t reserve_size = size * (1 + 1 + 1);
+        if constexpr (is_binary_str)
+            reserve_size += reserve_size <= 0 ? data_from->getSizeForReserve() : size * flen;
+        else
+            reserve_size += data_from->getSizeForReserve();
+        JsonBinary::JsonBinaryWriteBuffer write_buffer(data_to, reserve_size);
+        ColumnString::Chars_t tmp_buf;
+        for (size_t i = 0; i < size; ++i)
+        {
+            const auto & slice = data_from->getWhole();
+            if constexpr (is_binary_str)
+            {
+                if constexpr (check_null_for_binary_str)
+                {
+                    if (null_map_from[i])
+                    {
+                        writeChar(0, write_buffer);
+                        offsets_to[i] = write_buffer.count();
+                        data_from->next();
+                        continue;
+                    }
+                }
+                if (unlikely(flen <= 0))
+                {
+                    JsonBinary::appendOpaque(
+                        write_buffer,
+                        JsonBinary::Opaque{from_type_code, StringRef{slice.data, slice.size}});
+                }
+                else
+                {
+                    auto size_t_flen = static_cast<size_t>(flen);
+                    if (slice.size >= size_t_flen)
+                    {
+                        JsonBinary::appendOpaque(
+                            write_buffer,
+                            JsonBinary::Opaque{from_type_code, StringRef{slice.data, size_t_flen}});
+                    }
+                    else
+                    {
+                        if (tmp_buf.size() < size_t_flen)
+                            tmp_buf.resize(size_t_flen);
+                        std::memcpy(tmp_buf.data(), slice.data, slice.size);
+                        std::fill(tmp_buf.data() + slice.size, tmp_buf.data() + size_t_flen, 0);
+                        JsonBinary::appendOpaque(
+                            write_buffer,
+                            JsonBinary::Opaque{from_type_code, StringRef{tmp_buf.data(), size_t_flen}});
+                    }
+                }
+            }
+            else
+            {
+                JsonBinary::appendOpaque(
+                    write_buffer,
+                    JsonBinary::Opaque{from_type_code, StringRef{slice.data, slice.size}});
+            }
+            writeChar(0, write_buffer);
+            offsets_to[i] = write_buffer.count();
+            data_from->next();
+        }
+    }
+
+    template <bool is_nullable>
+    static void doExecuteForParsingJson(
+        ColumnString::Chars_t & data_to,
+        ColumnString::Offsets & offsets_to,
+        const std::unique_ptr<IStringSource> & data_from,
+        const NullMap & null_map_from,
+        size_t size)
+    {
+        // json_type + string end char 0 + value
+        size_t reserve_size = (size * (1 + 1)) + data_from->getSizeForReserve();
```

Comment:
Ditto, I think for string input column, `data_from->getSizeForReserve()` already count the string end char 0

Reply:
yes, you are right, changed.